pkg/controller/backup_controller.go

```diff
@@ -668,3 +710,18 @@ func closeAndRemoveFile(file *os.File, log logrus.FieldLogger) {
 		log.WithError(err).WithField("file", file.Name()).Error("error removing file")
 	}
 }
+
+// encodeToJSONGzip takes arbitrary Go data and encodes it to GZip compressed JSON in a buffer, as well as a description of the data to put into an error should encoding fail.
+func encodeToJSONGzip(data interface{}, desc string) (*bytes.Buffer, error) {
+	buf := new(bytes.Buffer)
+	gzw := gzip.NewWriter(buf)
+
+	if err := json.NewEncoder(gzw).Encode(data); err != nil {
+		return nil, errors.Wrapf(err, "error encoding %s", desc)
```

Comment:
if we hit this condition, then we never close `gzw` - probably just close it explicitly in here. the challenge is always what to do if *that* returns an error -- one option would be to pass a logger into this function and just log it, and return the encoding error.

Reply:
We could also close it with a defer and wrap with a function so we can grab the error, and used a named return on the error to get the error returned out instead of logged... I'm not a huge fan of named returns, but in this case it means both errors would be returned instead of one returned and one logged.

Looking at how the returned errors from the `persistBackup` function are used, they're all fatal, not logged, so I'm leaning towards trying to find a way to return them.

In code, this would be like

```go
defer func () {
}()
```

Reply:
Doing it this way may also require some more wrapping, because we don't want the defer's error overwriting any encoding errors, either.

Reply:
yeah, you can use that approach too - I usually try to avoid it because it gets complicated quickly. A third option would be to return a `[]error`, and append errors to that as they're encountered. Since the calling function also returns a `[]error`, that wouldn't be too bade.

Reply:
👍 That makes sense, especially since that's what the code originally did anyway. It took all these encoding & closing errors and put them in a slice, so I'll just do the same.

Reply:
sgtm!